cocotb/types/array.py

```diff
@@ -247,18 +296,20 @@ def __getitem__(self, item):
             "indexes must be ints or slices, not {}".format(type(item).__name__)
         )
 
-    @overload
-    def __setitem__(self, item: int, value: Any) -> None:
+    @typing.overload
+    def __setitem__(self, item: int, value: T) -> None:
         pass  # pragma: no cover
 
-    @overload
-    def __setitem__(self, item: slice, value: Iterable[Any]) -> None:
+    @typing.overload
+    def __setitem__(self, item: slice, value: typing.Iterable[T]) -> None:
         pass  # pragma: no cover
 
-    def __setitem__(self, item, value):
+    def __setitem__(
+        self, item: typing.Union[int, slice], value: typing.Union[T, typing.Iterable[T]]
+    ) -> None:
         if isinstance(item, int):
             idx = self._translate_index(item)
-            self._value[idx] = value
+            self._value[idx] = typing.cast(T, value)
```

Comment:
I'm surprised I've not seen these casts needed before, and that the mypy docs don't mention the need for them. At any rate, they're obviously at worst only unecessary anyway, and I'll take your word for it that mypy requires them.

Reply:
In this instance it's necessary because `value` is `typing.Union[T, typing.Iterable[T]]` at this point. You need to narrow it somehow. `item` is already narrowed here to `int` from the `isinstance` checks.